pkg/apis/k0s/v1beta1/cplb.go

```diff
@@ -79,52 +102,185 @@ type VRRPInstance struct {
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Maximum=2678400
+	// +kubebuilder:default=360
+	// +optional
+	PersistenceTimeoutSeconds int `json:"persistenceTimeoutSeconds,omitempty"`
+}
+
+// KeepalivedLBAlgo describes the load balancing algorithm.
+// +kubebuilder:validation:Enum=rr;wrr;lc;wlc;lblc;dh;sh;sed;nq
+type KeepalivedLBAlgo string
+
+const (
+	RRAlgo   KeepalivedLBAlgo = "rr"
+	WRRAlgo  KeepalivedLBAlgo = "wrr"
+	LCAlgo   KeepalivedLBAlgo = "lc"
+	WLCAlgo  KeepalivedLBAlgo = "wlc"
+	LBLCAlgo KeepalivedLBAlgo = "lblc"
+	DHAlgo   KeepalivedLBAlgo = "dh"
+	SHAlgo   KeepalivedLBAlgo = "sh"
+	SEDAlgo  KeepalivedLBAlgo = "sed"
+	NQAlgo   KeepalivedLBAlgo = "nq"
+)
+
+// KeepalivedLBKind describes the load balancing forwarding method.
+// +kubebuilder:validation:Enum=NAT;DR;TUN
+type KeepalivedLBKind string
+
+const (
+	NATLBKind KeepalivedLBKind = "NAT"
+	DRLBKind  KeepalivedLBKind = "DR"
+	TUNLBKind KeepalivedLBKind = "TUN"
+)
+
+// validateVirtualServers validates existing configuration and sets the default
+// values of undefined fields.
+func (k *KeepalivedSpec) validateVirtualServers() []error {
+	errs := []error{}
+	for i := range k.VirtualServers {
+		if k.VirtualServers[i].IPAddress == "" {
+			errs = append(errs, errors.New("IPAddress must be defined"))
+		}
+		if net.ParseIP(k.VirtualServers[i].IPAddress) == nil {
+			errs = append(errs, fmt.Errorf("invalid IP address: %s", k.VirtualServers[i].IPAddress))
+		}
+
+		if k.VirtualServers[i].LBAlgo == "" {
+			k.VirtualServers[i].LBAlgo = RRAlgo
+		} else {
+			switch k.VirtualServers[i].LBAlgo {
+			case RRAlgo, WRRAlgo, LCAlgo, WLCAlgo, LBLCAlgo, DHAlgo, SHAlgo, SEDAlgo, NQAlgo:
+				// valid LBAlgo
+			default:
+				errs = append(errs, fmt.Errorf("invalid LBAlgo: %s ", k.VirtualServers[i].LBAlgo))
+			}
+		}
+
+		if k.VirtualServers[i].LBKind == "" {
+			k.VirtualServers[i].LBKind = DRLBKind
+		} else {
+			switch k.VirtualServers[i].LBKind {
+			case NATLBKind, DRLBKind, TUNLBKind:
+				// valid LBKind
+			default:
+				errs = append(errs, fmt.Errorf("invalid LBKind: %s ", k.VirtualServers[i].LBKind))
+			}
+		}
+
+		if k.VirtualServers[i].PersistenceTimeoutSeconds == 0 {
+			k.VirtualServers[i].PersistenceTimeoutSeconds = 360
+		} else if k.VirtualServers[i].PersistenceTimeoutSeconds < 1 || k.VirtualServers[i].PersistenceTimeoutSeconds > 2678400 {
+			errs = append(errs, errors.New("PersistenceTimeout must be in the range of 1-2678400"))
+		}
+
+		k.VirtualServers[i].DelayLoop = metav1.Duration{Duration: k.VirtualServers[i].DelayLoop.Truncate(time.Microsecond)}
+		if k.VirtualServers[i].DelayLoop.Duration < time.Duration(0) {
+			errs = append(errs, errors.New("DelayLoop must be a positive"))
+		}
+	}
+	return errs
+}
+
+// Validate validates the ControlPlaneLoadBalancingSpec
+func (c *ControlPlaneLoadBalancingSpec) Validate(externalAddress string) []error {
+	if c == nil || !c.Enabled {
```

Comment:
We should validate the config in any case, even if it's disabled.

```suggestion
	if c == nil {
```